### PR TITLE
Stream support: VTP and VTU writing functions

### PR DESCRIPTION
**Pull request by maxGimeno**
## Summary of Changes

Add functions to write surface meshes, c3t3 and cdt to vtp and vtu format, without using the VTK lib, which avoids having to use intermediary structures.
## Release Management

* Affected package(s):Mesh_2, Mesh_3
* Small feature: https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/2D_and_3D_Mesh_Generation/VTK_output
* Pre-approval date: 2018/01/17

### CODE REVIEW

**Comment by sloriot** [comment]
I don't see any `CMakeLists.txt` update (in particular in the Polyhedron demo). Is this expected?

**Comment by maxGimeno** [comment]
Yes. It only affects the reading, so we still need VTK to read the data. Do you want me to add #ifdef directives to filtrate what always work and what needs vtk ? It seems like a good idea nw that you mention it.

**Comment by sloriot** [comment]
If you are no longer using data structure from VTK, I don't see why VTK would be needed at all. So `find_package(VTK)` could be removed from some plugins.

**Comment by maxGimeno** [comment]
> It only affects the writing

We still use vtk structures for reading data, so we still need VTK libs.

**Comment by lrineau** [comment]
The idea is that writing a VTU/VTP file is not too complicated, because we master everything, and we can chose a particular file format, in the set of things allowed by the VTK file format description.

But reading VTU/VTP files is another story: one has to be able to read all possible valid VTU/VTP files. That includes at least a full XML parser, the possibility to decode base64, gzip compression, and maybe other things.

So we will be:
  - very strict in out export format, writing VTU/VTP file in a specific sub-format, with a code that does not use the VTK libraries,
  - and we have to be conservative in our import of VTK files, and that requires VTK libraries.

**Comment by lrineau** [comment]
Travis reports a license issue:
```
=> There are modifications in licenses:
 M Stream_support/package_info/Stream_support/license.txt
```
https://travis-ci.org/CGAL/cgal/jobs/421126628#L1867

**Comment by lrineau** [comment]
https://travis-ci.org/CGAL/cgal/jobs/436537979#L3618
```
[ 93%] Building CXX object CMakeFiles/Constrained_Delaunay_triangulation_2.dir/Constrained_Delaunay_triangulation_2_autogen/mocs_compilation.cpp.o
/home/travis/build/CGAL/cgal/GraphicsView/demo/Triangulation_2/Constrained_Delaunay_triangulation_2.cpp:25:10: fatal error: 
      'CGAL/IO/vtk_io_2.h' file not found
#include <CGAL/IO/vtk_io_2.h>
         ^
1 error generated.
make[2]: *** [CMakeFiles/Constrained_Delaunay_triangulation_2.dir/Constrained_Delaunay_triangulation_2.cpp.o] Error 1
make[1]: *** [CMakeFiles/Constrained_Delaunay_triangulation_2.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by lrineau** [comment]
We are close:
```
.. Checking SPDX license identifier presence in header files...
The following files do not have a SPDX license identifier:
Mesh_2/include/CGAL/IO/vtk_io_2.h
Mesh_3/include/CGAL/IO/vtk_io.h
```
https://travis-ci.org/CGAL/cgal/jobs/437101040#L1815

**Comment by lrineau** [comment]
```
.. Checking include directives in GPL header files...
The following files do not have an include directive of the package license header:
Mesh_2/include/CGAL/IO/vtk_io_2.h
Mesh_3/include/CGAL/IO/vtk_io.h
```

**Comment by lrineau** [comment]
@maxGimeno: why the `TODO` label?

**Comment by lrineau** [comment]
There seem to be indentation issues in this PR.

**Comment by maxGimeno** [comment]
> @maxGimeno: why the `TODO` label?

because of 223fa68 I guess, it says "missing doc"

**Comment by lrineau** [comment]
TODO: review the documentation (small feature)?

File: Mesh_2/include/CGAL/IO/vtk_io_2.h [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+//! 
+//! \tparam CDT a `Constrained_Delaunay_triangulation_2`.
+//! 
+//! \param os a `std::ostream`.
+//! \param cdt an instance of `CDT` to be written.
+//! \param binary decides if the data should be written in binary(`true`)
+//!   or in ASCII(`false`).
+//!
+template <class CDT>
+void write_unstructured_grid_2(std::ostream& os,
```
Comment (lrineau) [inline Mesh_2/include/CGAL/IO/vtk_io_2.h]: I think the export of a `CDT_2` to a VTU file should also output the constrained edges, if possible. It seems the [`vtkUnstructuredGrid`] can mix type of "cells".

[`vtkUnstructuredGrid`]: https://www.vtk.org/doc/nightly/html/classvtkUnstructuredGrid.html

**Comment by lrineau** [comment]
@maxGimeno You must now create the corresponding small-feature.

**Comment by sloriot** [comment]
@lrineau would it make sense to move and document `vtk_io.h` in `Stream_support`?

**Comment by sloriot** [comment]
Why is `write_VTP` have a binary option that `write_VTU` does not have? Note that you can also write in XML format for polydata files.
It might be that you only have one implementation but I wonder if the name of the function shouldn't be different if you have XML vs _regular_ writing.

File: Polyhedron/demo/Polyhedron/Plugins/IO/VTK_io_plugin.cpp [hunk 2]
```diff
@@ -163,6 +173,7 @@ namespace CGAL{
       vtkCell* cell_ptr = poly_data->GetCell(i);
 
       vtkIdType nb_vertices = cell_ptr->GetNumberOfPoints();
+      std::cout<<nb_vertices<<std::endl;
```
Comment (sloriot) [inline Polyhedron/demo/Polyhedron/Plugins/IO/VTK_io_plugin.cpp]: remove me

**Comment by lrineau** [comment]
[**@sloriot**](https://github.com/sloriot) commented on [Nov 14, 2018, 7:12 PM GMT+1](https://github.com/CGAL/cgal/pull/3297#issuecomment-438762493 "2018-11-14T18:12:38Z - Replied by Github Reply Comments"):
> @lrineau would it make sense to move and document `vtk_io.h` in `Stream_support`?

We do not want `Stream_support` to depend on GPL packages. That is why the functionality has to be split in the different packages owning the data structures.

**Comment by sloriot** [comment]
Then move the code to `Polyhedron_IO` and the doc [here](https://cgal.geometryfactory.com/CGAL/doc/master/BGL/group__PkgBGLIOFct.html).

**Comment by maxGimeno** [comment]
@sloriot so more splitting and more duplication then ?

**Comment by lrineau** [comment]
By the way, there is also a conflict with `master`.

**Comment by sloriot** [comment]
Think as a user: where do you expect to find a function writing a TriangleMesh? with functions to write an OFF/STL or in mesh_3?

**Comment by sloriot** [comment]
Note that I don't think you have to duplicate the doc. You can use input file from other packages. If you don't find the macro specific for each package name, tell me and I'll look.

**Comment by lrineau** [comment]
I agree: `CGAL::write_VTP` does not deal with Mesh_3 at least. It should be with the output functions of the BGL.

**Comment by afabri** [comment]
You can also add a `\sa` and it even might make sense to add examples in the chapters where it can be used and to have these examples in the User Manual,

**Comment by sloriot** [comment]
Simply move them all to Stream_support. Nothing prevents you to add a link to `write_VTU` in the classified reference manual that will point to BGL if you want to. In the end, I hope we will have a proper IO package and all those functions will end there.

**Comment by maxGimeno** [comment]
So we want Stream_support to be GPL after all ?

**Comment by sloriot** [comment]
For now, let's make something consistent and put the code in `Polyhedron_IO` that is GPL and the doc in BGL.

**Comment by maxGimeno** [comment]
The question is about duplicating the shared code between write_VTU and write_VTP. If I duplicate the vtp (mesh) part is not GPL anymore, the GPL comes from the c3t3 stuff. And if I don't duplicate, I will not put functions that take triangulations into BGL nore Polyhedron_IO, it would make no sense.

**Comment by lrineau** [comment]
The 2D and 3D functions `write_VTU` have am ambiguity, when called with two arguments:
https://github.com/CGAL/cgal/blob/45b19c81349810c57bb9e114b84aa949faf5661b/Mesh_2/include/CGAL/IO/vtk_io_2.h#L277-L280
https://github.com/CGAL/cgal/blob/45b19c81349810c57bb9e114b84aa949faf5661b/Mesh_3/include/CGAL/IO/vtk_io.h#L521-L523

**Comment by lrineau** [comment]
[**@maxGimeno**](https://github.com/maxGimeno) commented on [Nov 20, 2018, 1:59 PM GMT+1](https://github.com/CGAL/cgal/pull/3297#issuecomment-440264865 "2018-11-20T12:59:34Z - Replied by Github Reply Comments"):
> The question is about duplicating the shared code between write_VTU and write_VTP. If I duplicate the vtp (mesh) part is not GPL anymore, the GPL comes from the c3t3 stuff. And if I don't duplicate, I will not put functions that take triangulations into BGL nore Polyhedron_IO, it would make no sense.

The common code, with `write_vector`, `write_points` could be in Stream_support, undocumented.

**Comment by sloriot** [comment]
> Why is `write_VTP` have a binary option that `write_VTU` does not have? Note that you can also write in XML format for polydata files.
> It might be that you only have one implementation but I wonder if the name of the function shouldn't be different if you have XML vs _regular_ writing.

Any comment on that?

**Comment by maxGimeno** [comment]
Yes, I must have removed the `binary` parameter while developping for removing the ambiguity between the 2 versions, and forgot about it at some point. 
>Note that you can also write in XML format for polydata files.
It might be that you only have one implementation but I wonder if the name of the function shouldn't be different if you have XML vs regular writing.

"regular writing" is not a thing in VTP, it is for legacy vtk format. XML is mandatory for all recent VTK formats.

File: Mesh_2/doc/Mesh_2/CGAL/IO/vtk_io_2.h [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+//! 
+//! \tparam CDT a `Constrained_Delaunay_triangulation_2`.
+//! 
+//! \param os a `std::ostream`.
+//! \param tr an instance of `CDT` to be written.
+//! \param binary decides if the data should be written in binary(`true`)
+//!   or in ASCII(`false`).
+//!
+template <class CDT>
+void write_VTU_2D(std::ostream& os,
```
Comment (sloriot) [inline Mesh_2/doc/Mesh_2/CGAL/IO/vtk_io_2.h]: I don't like this name, worse case to respect CGAL conventions it should be `_2` but for free function we usually don't have dimension suffix.

**Comment by afabri** [comment]
Note that we have `write_off()`, `write_obj()` .....  so maybe rename to `write_medit()`

File: Mesh_2/doc/Mesh_2/CGAL/IO/vtk_io_2.h (hunk 3 above)
Comment (sloriot) [reply Mesh_2/doc/Mesh_2/CGAL/IO/vtk_io_2.h]: + `changes.md` has not been updated.

**Comment by lrineau** [comment]
[**@sloriot**](https://github.com/sloriot) commented on [Nov 21, 2018, 4:07 PM GMT+1](https://github.com/CGAL/cgal/pull/3297#issuecomment-440696962 "2018-11-21T15:07:12Z - Replied by Github Reply Comments"):
> @lrineau shouldn't `write_VTU` for c3t3 be renamed `CGAL::output_to_VTU` to be consistent with `CGAL::output_to_medit()`?

Why not. The benefit would also be that it will remove the naming conflict with the 2D case,

**Comment by lrineau** [comment]
[**@afabri**](https://github.com/afabri) commented on [Nov 21, 2018, 4:08 PM GMT+1](https://github.com/CGAL/cgal/pull/3297#issuecomment-440697740 "2018-11-21T15:08:53Z - Replied by Github Reply Comments"):
> Note that we have `write_off()`, `write_obj()` .....  so maybe rename to `write_medit()`

If anybody "rename" that function, please keep a backward compatibility function, that forwards to the new name. Because that function is used a lot in our code base, outside CGAL.

**Comment by maxGimeno** [comment]
I vote for output_to_vtu(). write_medit would probably never be used because of legacy write_to_medit() which is used everywhere, and if it fixes the conflicts it's almost a sufficient reason in itself.

**Comment by lrineau** [comment]
There should be a second submission of the small feature, because almost everything has changed since the first one.

**Comment by maxGimeno** [comment]
vtp_io.h cannot go in BGL because of the use of the named parameters.

**Comment by maxGimeno** [comment]
@lrineau can I get an update on this ? you haven't validated the changes in the PR and the second Round still has no review in the SF.
@afabri you were the only reviewer of the first round, does this one satisfies you ?

**Comment by afabri** [comment]
> 
> 
> @lrineau can I get an update on this ? you haven't validated the changes in the PR and the second Round still has no review in the SF.
> @afabri you were the only reviewer of the first round, does this one satisfies you ?

I don't like that we have different names, only because the type of the input is a  template parameter. And if we keep different names they should not be different because we vary between `output_` and `write_`, but because we encode the input type in the name.

**Comment by maxGimeno** [comment]
@afabri I renamed the functions to include the input type

File: Installation/CHANGES.md [hunk 4]
```diff
@@ -64,9 +64,9 @@ Release date: March 2019
 
 ### 2D and 3D Mesh Generation
 -   Added 3 functions for writing in modern VTK formats:
-    -`CGAL::output_to_vtu()`
-    -`CGAL::write_VTU()`
-    -`CGAL::write_VTP()`
+    -`CGAL::write_C3T3_to_VTU()`
```
Comment (lrineau) [inline Installation/CHANGES.md]: Remove "modern" from the text. VTK XML formats date from VTK 4.0. It is more than 15 years old (2001-08-22).

If you want to qualify the format, you can say "in VTK XML formats".

File: Installation/CHANGES.md [hunk 5]
```diff
@@ -1576,14 +1576,14 @@ Release date: October 2014
 -   Changes in the set of supported platforms:
     -   The Microsoft Windows Visual C++ compiler 2008 (VC9) is no
         longer supported since CGAL-4.5.
--   Since CGAL version 4.0, Eigen was the recommended third-party
+-   Since CGAL version 4.0, Eigen was the recommended third-party
```
Comment (lrineau) [inline Installation/CHANGES.md]: Please revert all your changes to `CHANGES.md` that are not our WKT.

**Comment by maxGimeno** [comment]
@lrineau I think you can resolve your changes request  ?

File: BGL/doc/BGL/PackageDescription.txt [hunk 6]
```diff
@@ -708,6 +708,7 @@ user might encounter.
 ## I/O Functions ##
 - \link PkgBGLIOFct CGAL::read_off() \endlink
 - \link PkgBGLIOFct CGAL::write_off() \endlink
+- \link PkgBGLIOFct `CGAL::write_TM_to_VTP()` \endlink
```
Comment (lrineau) [inline BGL/doc/BGL/PackageDescription.txt]: Why are there quotes for `CGAL::write_TM_to_VTP()` here, but not for the other functions?

File: BGL/include/CGAL/IO/vtp_io.h [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+  std::size_t off = 0;
+  for( face_iterator fit = faces(mesh).begin() ;
+       fit != faces(mesh).end() ;
+       ++fit )
+    {
+      off += 3;
+      offsets.push_back(off);
+      BOOST_FOREACH(vertex_descriptor v,
+                    vertices_around_face(halfedge(*fit, mesh), mesh))
+	connectivity_table.push_back(V[v]);
```
Comment (lrineau) [inline BGL/include/CGAL/IO/vtp_io.h]: Indentation

File: BGL/include/CGAL/IO/vtp_io.h [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+    for( face_iterator fit = faces(mesh).begin() ;
+	 fit != faces(mesh).end() ;
+	 ++fit )
+      {
+	BOOST_FOREACH(vertex_descriptor v,
+                      vertices_around_face(halfedge(*fit, mesh), mesh))
+	  os << V[v] << " ";
+      }
+    os << "      </DataArray>\n";
+  }
```
Comment (lrineau) [inline BGL/include/CGAL/IO/vtp_io.h]: The indentation looks random.

File: BGL/include/CGAL/IO/vtp_io.h [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+    std::size_t polys_offset = 0;
+    for( face_iterator fit = faces(mesh).begin() ;
+	 fit != faces(mesh).end() ;
+	 ++fit )
+      {
+	polys_offset += 3;
+	os << polys_offset << " ";
+      }  
+    os << "      </DataArray>\n";
+  }
```
Comment (lrineau) [inline BGL/include/CGAL/IO/vtp_io.h]: Random indentation.

**Comment by maxGimeno** [comment]
These files were full of tabs :/

**Comment by sloriot** [comment]
Why did you change the names? @afabri did not like the previous name but they were consistent at the package level. The new names are ugly.

**Comment by sloriot** [comment]
\+ I don't think it is the goal of this PR to change the naming convention in Mesh_3.

**Comment by maxGimeno** [comment]
I asked @afabri for a review, if I shouldn't do what he tells me what's a reviewer needed for ?

**Comment by sloriot** [comment]
The discussion is getting too long on the PR but IIRC, @lrineau and I agreed on the previous names.

**Comment by lrineau** [comment]
The review and discussion about this small feature is very hard to follow, because all a lot of our review rules have been broken. I gather here all the discussion, so that we can conclude. That feature is important for a customer of ours.

It seems that the [first version of the PR was][1] (29 October 2018):
> -   Added 3 functions for writing in modern VTK formats:
>     - `CGAL::write_unstructured_grid_3(std::ostream& os, const C3T3& c3t3)`
>     - `CGAL::write_unstructured_grid_2(std::ostream& os, const CDT& tr, bool binary = true)`
>     - `CGAL::write_polydata(std::ostream& os, const TriangleMesh& mesh, bool binary, const NamedParameters& np)`

[1]: https://cgal.geometryfactory.com/CGAL/Members/w/index.php?title=Features/Small_Features/2D_and_3D_Mesh_Generation/VTK_output&oldid=39885#Submission

But then [Andreas said][review 1] (29 October 2018):
> - Wouldn't it make sense to have `VTU` in the function name, as we do for `OFF` or `OBJ`.
> - The documentation of the function can directly start with writes

[review 1]: https://cgal.geometryfactory.com/CGAL/Members/w/index.php?title=Features/Small_Features/2D_and_3D_Mesh_Generation/VTK_output&oldid=39886#Discussion

Then that [changed to][2], without opening a new round (29 October 2018):
> - `CGAL::write_VTU_2D`
> - `CGAL::write_VTP`
> - `CGAL::write_VTU_3D`


[2]: https://cgal.geometryfactory.com/CGAL/Members/w/index.php?title=Features/Small_Features/2D_and_3D_Mesh_Generation/VTK_output&oldid=39889#Submission

Then there was a lot of discussion, around https://github.com/CGAL/cgal/pull/3297#issuecomment-440696962, 21 Novemver 2018.

And then the submission [changed to][3], in a second round (29 November 2018):
> -   Added 3 functions for writing in modern VTK formats:
>     - `CGAL::output_to_vtu()`
>     - `CGAL::write_VTU()`
>     - `CGAL::write_VTP()`

[3]: https://cgal.geometryfactory.com/CGAL/Members/w/index.php?title=Features/Small_Features/2D_and_3D_Mesh_Generation/VTK_output&oldid=40059#Submission_2


[Then Andreas said](https://github.com/CGAL/cgal/pull/3297#issuecomment-452669116 "2019-01-09T11:47:26Z - Replied by Github Reply Comments"):
> > @lrineau can I get an update on this ? you haven't validated the changes in the PR and the second Round still has no review in the SF.
> > @afabri you were the only reviewer of the first round, does this one satisfies you ?
> 
> I don't like that we have different names, only because the type of the input is a  template parameter. And if we keep different names they should not be different because we vary between `output_` and `write_`, but because we encode the input type in the name.

And then again the naming changes, without a new round of submission, with commit https://github.com/CGAL/cgal/commit/d46ed84db0d8fecfc46b7559f26e45686c9b71a2:
>    -`CGAL::write_C3T3_to_VTU()`
>    -`CGAL::write_CDT_to_VTU()`
>    -`CGAL::write_TM_to_VTP()`

**Comment by lrineau** [comment]
[**@maxGimeno**](https://github.com/maxGimeno) commented on [Jan 15, 2019, 4:04 PM GMT+1](https://github.com/CGAL/cgal/pull/3297#issuecomment-454423732 "2019-01-15T15:04:48Z - Replied by Github Reply Comments"):
> I asked @afabri for a review, if I shouldn't do what he tells me what's a reviewer needed for ?


[**@sloriot**](https://github.com/sloriot) commented on [Jan 15, 2019, 4:08 PM GMT+1](https://github.com/CGAL/cgal/pull/3297#issuecomment-454425136 "2019-01-15T15:08:36Z - Replied by Github Reply Comments"):
> The discussion is getting too long on the PR but IIRC, @lrineau and I agreed on the previous names.

Indeed, what a reviewer says can contradict with other reviewers. And Sébastien and I do disagree with Andreas.

**Comment by lrineau** [comment]
IRL we agreed on version 3, even if the naming for outputs of Mesh_3 are pretty badly chosen.

**Comment by lrineau** [comment]
Version 3 is:
> >     - `CGAL::output_to_vtu()`
> >     - `CGAL::write_VTU()`
> >     - `CGAL::write_VTP()`

File: Installation/CHANGES.md [hunk 10]
```diff
@@ -62,6 +62,12 @@ Release date: March 2019
     `Arr_polyline_traits_2`, `Arr_polycurve_traits_2`, and
     `Arr_polycurve_basic_traits_2`.
 
+### 2D and 3D Mesh Generation
+-   Added 3 functions for writing in XML VTK formats:
+    - `CGAL::output_to_vtu()`
+    - `CGAL::write_VTU()`
+    - `CGAL::write_VTP()`
```
Comment (lrineau) [inline Installation/CHANGES.md]: This modification is wrong because `write_VTP` is added in the BGL package. I also think that the changelog should give a short sentence about the usage of each function.

**Comment by sloriot** [comment]
I agree for lower case (it is what the official doc is using). About the naming the file, I'd go for the name of the function.

**Comment by lrineau** [comment]
I will implement what I proposed in my review (https://github.com/CGAL/cgal/pull/3297#pullrequestreview-193161942). And as I do not like the Boolean parameters, I will turn it into a use of the `CGAL::IO::Mode` enum from `<CGAL/IO/io.h>` (with `BINARY` and `ASCII`).

File: BGL/doc/BGL/PackageDescription.txt [hunk 11]
```diff
@@ -708,7 +708,7 @@ user might encounter.
 ## I/O Functions ##
 - \link PkgBGLIOFct CGAL::read_off() \endlink
 - \link PkgBGLIOFct CGAL::write_off() \endlink
-- \link PkgBGLIOFct CGAL::write_VTP() \endlink
+- \link PkgBGLIOFct CGAL::write_vtp() \endlink
```
Comment (lrineau) [inline BGL/doc/BGL/PackageDescription.txt]: @sloriot Why do we use here 
```
- \link PkgBGLIOFct .. \endlink
```
instead of just:
```
- `CGAL::write_vtp`
```
?

Comment (sloriot) [reply BGL/doc/BGL/PackageDescription.txt]: It is needed read/write off because there are different overloads in different packages. For `write_vtp` it is not needed IMO.

**Comment by lrineau** [comment]
# Fourth submission

## `CHANGES.md`
(The entry about `write_wrl()` was already there.)

### 2D and 3D Mesh Generation

-   Added two functions for writing in XML VTK formats:
    - `CGAL::write_vtu()`, that writes a 2D mesh in a `.vtu` file,
    - `CGAL::output_to_vtu()`, that writes a 3D mesh in a `.vtu` file.

### CGAL and the Boost Graph Library (BGL)

-   Added function `write_wrl()` for writing into VRML 2.0 format.
-   Added functions `CGAL::write_vtp()` for writing a triangulated
      face graph in a `.vtp` file (XML VTK format).

## Documentation
- [`write_vtu`](https://cgal.geometryfactory.com/~lrineau/VTK_io_v4/Mesh_2/group__PkgMesh2IO.html)
- [`output_to_vtu`](https://cgal.geometryfactory.com/~lrineau/VTK_io_v4/Mesh_3/group__PkgMesh3IOFunctions.html#ga129e2d17d549300bc0812501ba4d29ee)
- [`write_vtp`](https://cgal.geometryfactory.com/~lrineau/VTK_io_v4/BGL/group__PkgBGLIOFct.html#gad08084cb8ae81460855a1dd8e21abbfc)

**Comment by lrineau** [comment]
@sloriot I obviously agree with mine own changes. Please review. Compared to my proposal at https://github.com/CGAL/cgal/pull/3297#pullrequestreview-193161942, the main modification is the parameter `bool binary = true` turned into `IO::Mode mode = IO::BINARY` (because I do not like Boolean parameters).

File: BGL/include/CGAL/boost/graph/io.h [hunk 12]
```diff
@@ -401,6 +402,284 @@ bool write_inp(std::ostream& os,
...
+    internal::write_vtp::write_polys_points(os,mesh, np);
+    internal::write_vtp::write_polys(os,mesh, np);
+  }
+  os << "</VTKFile>\n";
+}
+
+/*! \ingroup PkgBGLIOFct
+ *  Overload that implements the default arguments
+ *
+ *  Calls `write_vtp(os, mesh, mode, parameters::all_default())`.
```
Comment (sloriot) [inline BGL/include/CGAL/boost/graph/io.h]: we usually don't document those.

File: BGL/include/CGAL/boost/graph/io.h [hunk 13]
```diff
@@ -401,6 +402,284 @@ bool write_inp(std::ostream& os,
...
+ */
+template<class TriangleMesh>
+void write_vtp(std::ostream& os,
+                      const TriangleMesh& mesh,
+               IO::Mode mode = IO::BINARY)
```
Comment (sloriot) [inline BGL/include/CGAL/boost/graph/io.h]: actually mode should be a named parameter. I can do it's simpler for you. We just need a name for the named parameter (`use_binary_mode(false/true)`?)

Comment (lrineau) [reply BGL/include/CGAL/boost/graph/io.h]: Yes, you are right. That would be a simpler API. `use_binary_mode` seems fine for the name.

**Comment by afabri** [comment]
@sloriot don't we exagerate a little bit with the named parameters?

**Comment by lrineau** [comment]
# Fifth submission

This submission includes the changes of https://github.com/CGAL/cgal/pull/3297/commits/eadb08d976161381dd80a1d2b1984bd8ac4d77cf:
- fixed typo,
- use named parameters for binary, in `write_VTP`.

## `CHANGES.md`
(The entry about `write_wrl()` was already there.)

### 2D and 3D Mesh Generation

-   Added two functions for writing in XML VTK formats:
    - `CGAL::write_vtu()`, that writes a 2D mesh in a `.vtu` file,
    - `CGAL::output_to_vtu()`, that writes a 3D mesh in a `.vtu` file.

### CGAL and the Boost Graph Library (BGL)

-   Added function `write_wrl()` for writing into VRML 2.0 format.
-   Added functions `CGAL::write_vtp()` for writing a triangulated
      face graph in a `.vtp` file (XML VTK format).

## Documentation
- [`write_vtu`](https://cgal.geometryfactory.com/~lrineau/VTK_io_v5/Mesh_2/group__PkgMesh2IO.html)
- [`output_to_vtu`](https://cgal.geometryfactory.com/~lrineau/VTK_io_v5/Mesh_3/group__PkgMesh3IOFunctions.html#ga129e2d17d549300bc0812501ba4d29ee)
- [`write_vtp`](https://cgal.geometryfactory.com/~lrineau/VTK_io_v5/BGL/group__PkgBGLIOFct.html#ga055a2c0e4599fc8c7d74505d93bdf166)

**Comment by lrineau** [comment]
Houston, we have a problem.
https://travis-ci.org/CGAL/cgal/jobs/480980634#L7263
```
In file included from /home/travis/build/CGAL/cgal/dep_check_build/test_headers.cpp:431:
/home/travis/build/CGAL/cgal/Mesh_3/include/CGAL/IO/output_to_vtu.h:153:1: error: 
      redefinition of 'write_points_tag'
write_points_tag(std::ostream& os,
^
/home/travis/build/CGAL/cgal/Mesh_2/include/CGAL/IO/write_vtu.h:209:1: note: 
      previous definition is here
write_points_tag(std::ostream& os,
^
```

**Comment by maxGimeno** [comment]
Travis report : In file included from /home/travis/build/CGAL/cgal/Installation/../Point_set_processing_3/include/CGAL/OpenGR/register_point_sets.h:31:
/home/travis/build/CGAL/cgal/Point_set_processing_3/include/CGAL/OpenGR/compute_registration_transformation.h:33:10: fatal error: 'gr/algorithms/FunctorSuper4pcs.h' file not found
#include <gr/algorithms/FunctorSuper4pcs.h>
         ^

**Comment by lrineau** [comment]
[**@lrineau**](https://github.com/lrineau) commented on [Jan 18, 2019, 5:16 PM GMT+1](https://github.com/CGAL/cgal/pull/3297#issuecomment-455601515 "2019-01-18T16:16:49Z - Replied by Github Reply Comments"):
> Houston, we have a problem.
> https://travis-ci.org/CGAL/cgal/jobs/480980634#L7263
> ```
> In file included from /home/travis/build/CGAL/cgal/dep_check_build/test_headers.cpp:431:
> /home/travis/build/CGAL/cgal/Mesh_3/include/CGAL/IO/output_to_vtu.h:153:1: error: 
>       redefinition of 'write_points_tag'
> write_points_tag(std::ostream& os,
> ^
> /home/travis/build/CGAL/cgal/Mesh_2/include/CGAL/IO/write_vtu.h:209:1: note: 
>       previous definition is here
> write_points_tag(std::ostream& os,
> ^
> ```

I will deal with that error...

**Comment by lrineau** [comment]
I have reviewed the code of `output_to_vtu`, and there are two bugs:
  - far points are written in the file,
  - if the mode is `ASCII`, then the mesh domain indices are not filled before the call to `write_attribute_tag`, that needs a filled vector.

**Comment by lrineau** [comment]
Another bug:
https://github.com/CGAL/cgal/blob/ae108c5d96a00c33cb5778fb5bcd65278c4a1764/Mesh_3/include/CGAL/IO/output_to_vtu.h#L294-L295

According to the [documentation], the attributes of `CellData` can be:
> - `Scalars` — The name of the active scalars array, if any.
> - `Vectors` — The name of the active vectors array, if any.
> - `Normals` — The name of the active normals array, if any.
> - `Tensors` — The name of the active tensors array, if any.
> - `TCoords` — The name of the active texture coordinates array, if any.

[documentation]: http://vtk.org/VTK/img/file-formats.pdf

`Domain=` should be `Scalars=`.

**Comment by maxGimeno** [comment]
@lrineau I am pretty sure the only thing it would change is the name displayed in paraview. So if you want it to be called Scalars as in the convention of VTK, then it is a bug, but if you estimate that this property is the Domain as in CGAL, then it should remain "Domain".

**Comment by maxGimeno** [comment]
@lrineau Ok I think I got lost in the code, that's absolutely not referring to what I thougt, sorry.

**Comment by lrineau** [comment]
[**@lrineau**](https://github.com/lrineau) commented on [Jan 21, 2019, 5:13 PM GMT+1](https://github.com/CGAL/cgal/pull/3297/#issuecomment-456126158 "2019-01-21T16:13:43Z - Replied by Github Reply Comments"):
> I have reviewed the code of `output_to_vtu`, and there are two bugs:
>   - far points are written in the file,

Fixed in https://github.com/CGAL/cgal/pull/3297/commits/4d3df85cb3b02139b83ca27a4ee97f46f0d96b67.


[**@lrineau**](https://github.com/lrineau) commented on [Jan 21, 2019, 5:19 PM GMT+1](https://github.com/CGAL/cgal/pull/3297/#issuecomment-456127902 "2019-01-21T16:19:15Z - Replied by Github Reply Comments"):
> `Domain=` should be `Scalars=`.

Fixed in https://github.com/CGAL/cgal/pull/3297/commits/b74753680bf26253c2f9beaaaedd613b0e81c273.

The bug about `mids` and `ASCII` remains.

**Comment by lrineau** [comment]
[**@lrineau**](https://github.com/lrineau) commented on [Jan 22, 2019, 12:08 PM GMT+1](https://github.com/CGAL/cgal/pull/3297/#issuecomment-456361239 "2019-01-22T11:08:22Z - Replied by Github Reply Comments"):
> The bug about `mids` and `ASCII` remains.

Fixed in https://github.com/CGAL/cgal/pull/3297/commits/a57db0bac4cfffb97f421025de17996f2cc74d29.

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-105.shtml.

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-116.shtml
